packages/nitro-protocol/contracts/AssetHolder.sol

```diff
@@ -139,203 +139,120 @@ contract AssetHolder is IAssetHolder {
-        for (i = 0; i < allocation.length; i++) {
-            if (balance == 0) {
-                revert('_transfer | fromChannel affords 0 for destination');
-            }
-            uint256 _amount = allocation[i].amount;
-            if (allocation[i].destination == destination) {
-                if (balance < _amount) {
-                    affordsForDestination = balance;
-                    residualAllocationAmount = _amount - balance;
-                    balance = 0;
-                } else {
-                    affordsForDestination = _amount;
-                    residualAllocationAmount = 0;
-                    balance = balance.sub(_amount);
-                }
-                break; // means that i holds the index of the destination that may need to be altered or removed
-            }
-            if (balance < _amount) {
-                balance = 0;
-            } else {
-                balance = balance.sub(_amount);
-            }
-        }
-
-        require(affordsForDestination > 0, '_transfer fromChannel allocates 0 to destination');
-
-        // effects
-        holdings[fromChannelId] -= affordsForDestination;
-
-        // construct new outcome
+        (
+            Outcome.AllocationItem[] memory newAllocation,
+            bool safeToDelete,
+            uint256[] memory payouts,
+            uint256 totalPayouts
+        ) = _computeNewAllocation(initialHoldings, allocation, indices);
 
-        bytes memory encodedAllocation;
+        // *******
+        // EFFECTS
+        // *******
 
-        if (residualAllocationAmount > 0) {
-            // new allocation identical save for a single entry
-            Outcome.AllocationItem[] memory newAllocation = new Outcome.AllocationItem[](
-                allocation.length
-            );
-            for (uint256 k = 0; k < allocation.length; k++) {
-                newAllocation[k] = allocation[k];
-                if (allocation[k].destination == destination) {
-                    newAllocation[k].amount = residualAllocationAmount;
-                }
-            }
-            encodedAllocation = abi.encode(newAllocation);
-        }
+        uint256 newHoldings = initialHoldings.sub(totalPayouts);
 
-        if (residualAllocationAmount == 0) {
-            Outcome.AllocationItem[] memory splicedAllocation = new Outcome.AllocationItem[](
-                allocation.length - 1
-            );
-            // full payout so we want to splice a shorter outcome
-            for (uint256 k = 0; k < i; k++) {
-                splicedAllocation[k] = allocation[k];
-            }
-            for (uint256 k = i + 1; k < allocation.length; k++) {
-                splicedAllocation[k - 1] = allocation[k];
-            }
-            if (splicedAllocation.length == 0) {
-                deleteHash = true;
-            }
-            encodedAllocation = abi.encode(splicedAllocation);
+        if (newHoldings == 0) {
+            delete holdings[fromChannelId];
```

Comment:
Considering that mapping values default to `0`, does setting `holdings[fromChannelId] = 0` have a different effect than `delete holdings[fromChannelId]`?

Reply:
I don't think it has a different effect. So yeah we don't need this branching logic here.

Reply:
I suppose it's important to know for sure, since we wait to provide the gas refund is possible.

Reply:
The [docs ](https://docs.soliditylang.org/en/v0.7.5/types.html?highlight=delete#delete)are pretty clear 
> delete a assigns the initial value for the type to a. I.e. for integers it is equivalent to a = 0, 

and

> It is important to note that delete a really behaves like an assignment to a, i.e. it stores a new object in a.

Reply:
d867e74